core/src/display_object/video.rs

```diff
@@ -0,0 +1,336 @@
+//! Video player display object
+
+use crate::avm1::{Object as Avm1Object, StageObject as Avm1StageObject};
+use crate::avm2::{Object as Avm2Object, StageObject as Avm2StageObject};
+use crate::backend::render::BitmapHandle;
+use crate::backend::video::{EncodedFrame, FrameDependency, VideoStreamHandle};
+use crate::bounding_box::BoundingBox;
+use crate::collect::CollectWrapper;
+use crate::context::{RenderContext, UpdateContext};
+use crate::display_object::{DisplayObjectBase, TDisplayObject};
+use crate::prelude::*;
+use crate::tag_utils::{SwfMovie, SwfSlice};
+use crate::types::{Degrees, Percent};
+use crate::vminterface::{AvmObject, AvmType, Instantiator};
+use gc_arena::{Collect, GcCell, MutationContext};
+use std::borrow::{Borrow, BorrowMut};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+use swf::{CharacterId, DefineVideoStream, VideoFrame};
+
+/// A Video display object is a high-level interface to a video player.
+///
+/// Video data may be embedded within a variety of container formats, including
+/// a host SWF, or an externally-loaded FLV or F4V file. In the latter form,
+/// video framerates are (supposedly) permitted to differ from the stage
+/// framerate.
+#[derive(Clone, Debug, Collect, Copy)]
+#[collect(no_drop)]
+pub struct Video<'gc>(GcCell<'gc, VideoData<'gc>>);
+
+#[derive(Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct VideoData<'gc> {
+    base: DisplayObjectBase<'gc>,
+
+    /// The source of the video data (e.g. an external file, a SWF bitstream)
+    source: GcCell<'gc, VideoSource>,
+
+    /// The decoder stream that this video source is associated to.
+    stream: Option<CollectWrapper<VideoStreamHandle>>,
+
+    /// The last decoded frame in the video stream.
+    decoded_frame: Option<(u32, CollectWrapper<BitmapHandle>)>,
+
+    /// AVM representation of this video player.
+    object: Option<AvmObject<'gc>>,
+
+    /// List of frames which can be independently seeked to.
+    ///
+    /// Frames outside of this set must be decoded by playing each frame from
+    /// the last keyframe in order. Any out-of-order seeking will be snapped to
+    /// the prior keyframe. The first frame in the stream will always be
+    /// treated as a keyframe regardless of it being flagged as one.
+    keyframes: BTreeSet<u32>,
+}
+
+#[derive(Clone, Debug, Collect)]
+#[collect(require_static)]
+pub enum VideoSource {
+    SWF {
+        /// The movie that defined this video stream.
+        movie: Arc<SwfMovie>,
+
+        /// The video stream definition.
+        streamdef: DefineVideoStream,
+
+        /// The H.263 bitstream indexed by video frame ID.
```

Comment:
This isn't supposed to be codec-specific, right?

Reply:
Nope.